Add a way to create Music Discs without requiring a SoundEvent

TheSilkMiner opened the pull request:
### Abstract
Creating a `MusicDiscItem` requires the user to supply a `SoundEvent`, which is a `ForgeRegistryEntry`. For this reason, these entries should only be created and registered in their `RegistryEvent.Register`. In this case, though, the registry event for `SoundEvent`s fires after the one for `Item`s, making it effectively impossible to create a `MusicDiscItem`.

### The Changes
What this PR does is change this requirement so that the user can use a `Supplier<SoundEvent>`, so that the actual `SoundEvent` can be retrieved and loaded as needed. Vanilla has also been patched to account for this behavior.
Additionally, `WorldRenderer` has been patched not to be stupid, rendering the `RECORDS` map effectively useless, since the look up is now performed only if the two-parameter version of `playRecord` is called with a non-`null` `SoundEvent`, which vanilla doesn't call anymore. In fact, the `MusicDiscItem` that vanilla retrieves from `playEvent` isn't discarded anymore, but it's passed to Forge's three-parameter version.
This PR also provides a test mod that demonstrates the usage of the newly added hook.

TheSilkMiner committed:
Patch Music Disc not to require a SoundEvent for construction

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Avoid changing the type of a vanilla field

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=6523) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=6523) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MinecraftForge/MinecraftForge?pullRequest=6523) it.</sub>

gigaherz commented on patches/minecraft/net/minecraft/item/MusicDiscItem.java.patch:
```diff
@@ -0,0 +1,60 @@
+ public class MusicDiscItem extends Item {
+    private static final Map<SoundEvent, MusicDiscItem> field_150928_b = Maps.newHashMap();
+    private final int field_195977_c;
+-   private final SoundEvent field_185076_b;
```
~~We don't remove vanilla fields. That's a major binary break, and wouldn't be wanted even if we weren't in the no-breaking-changes stage.~~

~~I believe that alone makes this PR in the current state a no-go.~~

EDIT: oops! I misread the change. it's not removing a field, it's just removing a `final`. This should be in an AT, not patch. My comment about what I think the right solution would be still stands, though.

TheSilkMiner replied:
This is now outdated, since the field isn't definalized anymore

gigaherz reviewed (changes requested):
See the comment on the code for why I think this PR can't be merged as-is.

I think based on other cases in which we have dealt with this, the right solution would be:
1. Deprecate the `sound` field, and the `RECORDS` map, and add a `Supplier<SoundType>` that can be filled with a `RegistryDelegate` or `RegistryObject`.
2. Provide an alternative constructor that takes a `Supplier<SoundType>`, for use by mods. This would leave the `sound` field null, and NOT add the item to the map at all.
3. Patch the existing vanilla constructor to assign the supplier from `sound.delegate`, but leaving the existing logic as-is.
4. Add a forge override of `WorldRenderer#playRecord` that accepts the item description as a third param, and patch WorldRenderer#playEvent[case 1010:] to actually USE the item that it already obtained.

You could look at the BucketItem for an example of this being done.

Note that this is a different case from #6299 because the `RECORDS` map is **only used to get the item from a SoundEvent that was itself obtained from the item....** (🤦‍♂️), while the EGGS map is used in multiple places, some of them a bit too early in the loading process.

TheSilkMiner committed:
Rework patches as suggested by code review

Vanilla logic is now mostly intact and doesn't require reworking of
fields and definalizing. WorldRenderer has been patched not to be stupid
and use the item it has already

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Rework test mod to account for changes in patches

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner commented:
Addressed issues in above comment, reworking the patches as suggested,
~~The changes caused by the additional three-parameters method are actually more invasive than I thought: what should I do about them, since it seems they mostly refer to name changes?~~ Fixed with `forge.exc`

pupnewfster commented on patches/minecraft/net/minecraft/client/renderer/WorldRenderer.java.patch:
```diff
@@ -85,7 +85,56 @@
+-      ISound isound = this.field_147593_P.get(p_184377_2_);
++   @Deprecated // Forge: use item aware function below
++   public void func_184377_a(@Nullable SoundEvent p_184377_1_, BlockPos p_184377_2_)
++   {
```
There is no reason to move the bracket onto a new line. (patches are supposed to be as "simple" as possible)

gigaherz reviewed (approved):
Looks good to me now. :)

TheSilkMiner committed:
Add test mod

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Reduced patch size

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=6523) <br/>All committers have signed the CLA.

LexManos commented on src/main/resources/forge.exc:
```diff
@@ -118,3 +120,10 @@ net/minecraft/world/storage/loot/LootEntryEmpty.<init>(II[Lnet/minecraft/world/s
 net/minecraft/world/storage/loot/LootEntryItem.<init>(Lnet/minecraft/item/Item;II[Lnet/minecraft/world/storage/loot/functions/LootFunction;[Lnet/minecraft/world/storage/loot/conditions/LootCondition;Ljava/lang/String;)V=|p_i46644_1_,p_i46644_2_,p_i46644_3_,p_i46644_4_,p_i46644_5_,entryName
 net/minecraft/world/storage/loot/LootEntryTable.<init>(Lnet/minecraft/util/ResourceLocation;II[Lnet/minecraft/world/storage/loot/conditions/LootCondition;Ljava/lang/String;)V=|p_i46639_1_,p_i46639_2_,p_i46639_3_,p_i46639_4_,entryName
 net/minecraft/world/storage/loot/LootPool.<init>([Lnet/minecraft/world/storage/loot/LootEntry;[Lnet/minecraft/world/storage/loot/conditions/LootCondition;Lnet/minecraft/world/storage/loot/RandomValueRange;Lnet/minecraft/world/storage/loot/RandomValueRange;Ljava/lang/String;)V=|p_i46643_1_,p_i46643_2_,p_i46643_3_,p_i46643_4_
+<<<<<<< HEAD
```
Merge error

LexManos commented on src/test/java/net/minecraftforge/debug/item/MusicDiscTest.java:
```diff
@@ -0,0 +1,36 @@
+        SOUND_EVENTS.register(modBus);
+    }
+
+    static
```
This is just ugly, Just have 
```
 private static final RegisteryObject<SoundEvent> TEST_SOUND = ... 
 private static final RegistryObject<Item> TEST_DISC = ...
```
At the top

LexManos commented on src/test/resources/assets/music_disc_test/sounds.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "test_sound_event": {
```
Data generators would be nice for all the jsons

TheSilkMiner replied:
Should I create a new data provider for this situation in this PR or should I open up a new one with the specified addition?

LexManos replied:
Make a new PR for the added data gen.

LexManos commented on src/test/resources/assets/music_disc_test/sounds.json:
```diff
@@ -0,0 +1,13 @@
+        "stream": true,
+        "__content": "First 10 seconds of 'The Ice Giants' by Kevin MacLeod",
+        "__link": "https://incompetech.filmmusic.io/song/5745-the-ice-giants",
+        "__license": "http://creativecommons.org/licenses/by/4.0/"
```
Better to just reuse a vanilla sound so we don't have to deal with having other people's licenses and data in our repo.

TheSilkMiner committed:
Fix merge and code style issues

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Run license check

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

LexManos committed:
Merge branch '1.15.x' into feature/music-disc-sound-delegate
